chore(aws): add first iteration on eksctl usage

Langleu opened the pull request:
## Description

This PR is just preliminary and targeting a separate branch, not main! It's just a PR to get a technical review.
This does not follow writing best practices and will be reworked in the future when creating a PR from `INFEX-6` to `main`.

<!-- This helps the reviewers by providing an overview of what to expect in the PR. Linking to an issue is preferred but not required. -->
<!-- Add an assignee and use component: labels for good hygiene! -->

## When should this change go live?

<!-- PRs merged go to stage.docs.camunda.io first and must be manually released to docs.camunda.io. -->
<!-- Help the DevEx team prioritize our work (reviews, merges, etc.) by opening PRs sooner. -->

- [ ] This change is not yet live and should not be merged until {ADD_DATE} (apply `hold` label or convert to draft PR)?
- [ ] There is no urgency with this change.
- [ ] This change or page is part of a marketing blog, conference talk, or something else on a schedule.
- [ ] This functionality is already available but undocumented.
- [ ] This is a bug fix or security concern.

## PR Checklist

<!-- Keep in mind, Camunda maintains 18 months of versions. Backporting your change or including it in multiple versions is common. -->

- [ ] I have added changes to the relevant `/versioned_docs` directory, or they are not for an **already released version**.
- [ ] I have added changes to the main `/docs` directory (aka `/next/`), or they are not for **future versions**.
- [ ] My changes require an [Engineering review](https://github.com/camunda/camunda-docs/blob/main/howtos/documentation-guidelines.md#review-process), and I've assigned an engineering manager or tech lead as a reviewer, or my changes do not require an Engineering review.
- [ ] My changes require a [technical writer review](https://github.com/camunda/camunda-docs/blob/main/howtos/documentation-guidelines.md#review-process), and I've assigned @christinaausley as a reviewer, or my changes do not require a technical writer review.

christinaausley commented:
@Langleu Please go ahead and add the appropriate reviewers once you are ready 👍

Langleu commented:
thanks @christinaausley, forgot to add them in here but tagged them in the source ticket.
The target branch is not `main`, therefore I don't aim to get technical writer review yet, same for the PR https://github.com/camunda/camunda-docs/pull/2825.

maxdanilov reviewed (changes requested):
Leaving some comments suggestions, mainly of the styling nature.

I didn't have time to test all the commands yet (stopped on the DB subnet creation), will add more feedback soon.

As a summary I can already say I like the approach this guide takes, it's straightforward and easy to follow.

One solution to improve UX I would consider thinking of is how to make the interpolation operations easier for the customers (e.g. replacing all <clusterName> etc. is tedious, especially when this values need to be extracted manually, like the VPC ID or the subnet IDs.

maxdanilov commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
````diff
@@ -0,0 +1,288 @@
+Where `cluster.yaml` contains the following contents.
+Before applying, please adjust the `<region>` and `<zone>` and rename the cluster name with the placeholder `<clusterName>` to your liking.
+
+```yaml
````
This would also create somewhat an excplicit interface for the user to customize the parameters of their cluster.

Langleu replied:
should be possible as Docusaurus allows writing React components, so we can have a bit more interactive and embedded UI.
I'd suggest creating a ticket so we don't lose track of the idea.

I'm not sure it's a good idea to go with any sort of script as it should be OS agnostic and Windows is quite common.
The CLIs are fine since they work the same way.

maxdanilov replied:
Fair point about Windows users. The variable substitution looks same in Powershell and Shell it [seems](https://learn.microsoft.com/en-us/powershell/scripting/learn/deep-dives/everything-about-string-substitutions?view=powershell-7.3#variable-substitution), maybe that just using the variables is sufficient.

On the cross-platform approach, we can also consider using Docusaurus content tabs and provide 2 flavours of commands for users to switch between (Unix vs Windows).

Langleu replied:
For the first iteration, I'd suggest we concentrate on UNIX with a disclaimer in the consideration that it can be easily adapted to Windows.
In a future iteration, we either go the react way or use tabs. Of which I would find react easier since we don't have to continuously test it on windows.

maxdanilov commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
```diff
@@ -0,0 +1,288 @@
+  --host=<dbHost> \
+  --username=<username> \
+  --port=5432 \
+  --dbname=postgres
```
shouldn't we use the `<database>` name instead?  (the one used in `aws rds create-db-cluster`)

Langleu replied:
`postgres` should always be available so it doesn't overcomplicate things.
Nevertheless good point to have the value in sync with the C8 deployment

Langleu replied:
I'd still default to `postgres`, it will always be part of your created DB.

maxdanilov commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
````diff
@@ -0,0 +1,288 @@
+
+```shell
+psql \
+  --host=<dbHost> \
````
I think it should document how to define `dbHost` (which one to use - cluster (RW and/or RW) and/or the instance one?)

Langleu replied:
Added aws cli commands to automatically retrieve the value, obviously of the writer instance.
Also added environment variables to the kubectl run command to forward the values to the created pod for easier consumption.

Langleu commented:
good point on the interpolation topics!
For the placeholders, we could resolve to sed or sorts as a recommendation and example or trying it out with env vars, whether they trigger.
For figuring out the VPC or other IDs, I'll double check again whether there's a better solution. AWS CLI has a query option that could solve it already, alternative resolve to something like `jq`, but I think the query option is capable of doing the same. This would allow to have a single command and export it as environment variable.

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
```diff
@@ -0,0 +1,288 @@
+
+### Eksctl Prerequisites
+
+Setup authentication to allow interacting with `AWS` via the [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html) to configure access.
```
https://docs.aws.amazon.com/cli/latest/userguide/getting-started-quickstart.html

might be overwhelming but in details covers any case.

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
```diff
@@ -0,0 +1,288 @@
+
+Setup authentication to allow interacting with `AWS` via the [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html) to configure access.
+
+General advice, the user creating the resources will be the owner of those and always linked to those. Meaning the user will always have admin access on Kubernetes, except you delete it. Therefore, it can make sense to create an extra [IAM user](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users.html) that's solely used for the EKSCTL purposes.
```
Switched the link for https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users_create.html

We can link to access keys - https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_access-keys.html sadly they don't work a lot with IDs so we can't have good direct links and mention exporting env variables to use them.

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
```diff
@@ -0,0 +1,288 @@
+  ipFamily: IPv4
+managedNodeGroups:
+  - amiFamily: AmazonLinux2
+    desiredCapacity: 4
```
according to all the platform guides yes
https://docs.camunda.io/docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/

3 nodes alone are required because of 3 default Zeebe replicas and they have constraints to not spawn on the same node. I'd assume 1 additional node for misc.

With a bit of config adjustments one could probably run it with less as well.

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
````diff
@@ -0,0 +1,288 @@
+ eksctl create iamidentitymapping --cluster  <clusterName> --region=<region> --arn arn:aws:iam::123456:role/testing --group system:masters --username admin
+```
+
+Where `arn` is the `arn` of your user or the role. The `group` is the Kubernetes rule, where `system:masters` is equivalent to an admin role. Lastly `username` is either the username itself or the role name, which is used for logs.
````
yes can be anything, it's just for logging to identify who did what.
It's a mix of username + some invisibly added session marker.

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
````diff
@@ -0,0 +1,288 @@
+1. Figure out the VPC that is associated with the cluster
+   - Either via the [AWS UI](https://console.aws.amazon.com/eks/home)
+   - Or CLI - `aws ec2 describe-vpcs` and look for the VPC that contains the tag
+   ```yaml
````
I'll look into other options here again. AWS has a query option that should be able to do the same as jq.

Langleu replied:
redone it using the inbuilt query option

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
```diff
@@ -0,0 +1,288 @@
+aws rds create-db-subnet-group \
+    --db-subnet-group-name camunda-postgres \
+    --db-subnet-group-description "Subnet for Camunda Postgres" \
+    --subnet-ids subnet-0659823f3a8fa383c subnet-0f435ed77e62a13be subnet-06de5f9aa67849b1c
```
you're right, will rework to retrieve them.

Langleu replied:
reworked it and retrieved them via the aws cli and using filters and queries to automatically cut the right values.

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
```diff
@@ -0,0 +1,288 @@
+aws rds create-db-cluster \
+    --db-cluster-identifier <dbCluster> \
+    --engine aurora-postgresql \
+    --engine-version 15.2 \
```
you mean just `15.x` or `15.4`?
It will eventually be outdated anyway. I'll add a link to the supported environments https://docs.camunda.io/docs/reference/supported-environments/#camunda-8-self-managed

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
```diff
@@ -0,0 +1,288 @@
+- an AWS EKS Kubernetes Cluster with 4 nodes and the possibility to scale up further.
+- the [EBS CSI driver](https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html) is installed and configured.
+- a [managed PostgreSQL](https://aws.amazon.com/rds/postgresql/) instance is spawned that will be used by the Camunda applications.
+
```
no we can add that option with IRSA but you'll always have a username/password with Aurora, so you either use that or create separate accounts for Camunda purposes.
I've to look up that stuff in the docs of the products as not everyone is capable of creating the db but require manual work.

So far there's nothing 8.4+ specific in the setup.

maxdanilov replied:
got it, my bad. The guide is for Aurora flavour, does it make any difference if we use Aurora Postgres vs usual RDS Postgres? e.g. in how IRSA would work?

Langleu replied:
IRSA works for both aurora and classic postgres DBs.
The only difference is the way aurora and classic work, some are more suitable for A and others for B.
I think Keycloak would be fine with a classic postgres, web-modeler probably as well depends on the target user.

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
````diff
@@ -0,0 +1,288 @@
+3. Use eksctl to enable secret encryption
+
+```shell
+eksctl utils enable-secrets-encryption --cluster=<clusterName> --key-arn=<arnKey> --region=<region>
````
agree

Langleu replied:
reworked to environment variables and as a first step before creating the cluster to reuse it during cluster creation.

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
````diff
@@ -0,0 +1,288 @@
+General advice, the user creating the resources will be the owner of those and always linked to those. Meaning the user will always have admin access on Kubernetes, except you delete it. Therefore, it can make sense to create an extra [IAM user](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users.html) that's solely used for the EKSCTL purposes.
+
+<!--
+```shell
````
yes, excluded it from being rendered. It was the command to create the initial eksctl yaml.

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
````diff
@@ -0,0 +1,288 @@
+3. Use eksctl to enable secret encryption
+
+```shell
+eksctl utils enable-secrets-encryption --cluster=<clusterName> --key-arn=<arnKey> --region=<region>
````
will check it out, but still requires the previous creation of the key.

maxdanilov replied:
sure, the key could be made a prerequisite for the guide

Langleu replied:
done, worked wonderfully, thanks for the suggestion!

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
````diff
@@ -0,0 +1,288 @@
+
+```shell
+aws rds create-db-cluster \
+    --db-cluster-identifier <dbCluster> \
````
reworked it by initially defining the values so we can rely on environment variables.

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
```diff
@@ -0,0 +1,288 @@
+    --engine-version 15.2 \
+    --master-username <username> \
+    --master-user-password <password> \
+    --vpc-security-group-ids sg-004e773e9ac8369a9 \
```
replaced with environment variables that is automatically seeded a couple steps earlier.

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
```diff
@@ -0,0 +1,288 @@
+    --master-user-password <password> \
+    --vpc-security-group-ids sg-004e773e9ac8369a9 \
+    --availability-zones <zones> \ # eu-central-1a eu-central-1b
+    --database-name <dbName> \
```
I set the cluster and instance name to the same value.
DB Name is defaulted to `camunda` for the helm chart usage but can be overwritten if wanted.

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
````diff
@@ -0,0 +1,288 @@
+
+```shell
+aws rds create-db-instance \
+    --db-instance-identifier <dbInstance> \
````
I set both as the same value making those things easier since the aws commands always either target instance or cluster.

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
````diff
@@ -0,0 +1,288 @@
+3. Use eksctl to enable secret encryption
+
+```shell
+eksctl utils enable-secrets-encryption --cluster=<clusterName> --key-arn=<arnKey> --region=<region>
````
obsolete as now part of the initial cluster creation, saving the time.

Langleu commented on docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/eksctl.md:
```diff
@@ -0,0 +1,288 @@
+
+Following this guide will result in
+
+- an AWS EKS Kubernetes Cluster with 4 nodes and the possibility to scale up further.
```
added the versions + added versions to the used CLIs

Langleu commented:
- removed patch versions
- added postgres as variable and adjusted in the code blocks
- mentioned IAM being optional and made 2 collapsible blocks for cluster access and IAM setup
- made AWS org an environment variable and retrieved it via code
- cert-manager/external-dns mentioned being IAM setup only and added reference to Helm guide for the actual installation
- moved everything under C8 prerequisites
